lib/incito-browser/incito.js

```diff
@@ -1,308 +1,614 @@
-        }
-    ) {
+    constructor(containerEl, {incito = {}}) {
         this.containerEl = containerEl;
         this.incito = incito;
-        this.renderLaziness = renderLaziness;
         this.el = document.createElement('div');
+        this.imageCount = 0;
+        this.preloadImageCount = 20;
         this.ids = {};
-        this.views = flattenViews([], this.incito.root_view);
-        this.viewsLength = this.views.length;
-        this.viewIndex = 0;
-        this.lazyloadables = [];
-        this.lazyloader = throttle(this.lazyload.bind(this), 150);
-        this.renderedOutsideOfViewport = false;
-        this._events = {};
     }
 
-    bind(event, fn) {
-        this._events[event] = this._events[event] || [];
-        return this._events[event].push(fn);
-    }
+    start() {
+        const html = this.renderHtml();
+        const theme = this.incito.theme || {};
 
-    unbind(event, fn) {
-        if (this._events[event]) {
-            return this._events[event].splice(
-                this._events[event].indexOf(fn),
-                1
-            );
+        if (Array.isArray(theme.font_family)) {
+            this.el.style.fontFamily = theme.font_family.join(', ');
         }
-    }
 
-    trigger(event) {
-        if (this._events[event]) {
-            return this._events[event].map(function (e) {
-                return e.apply(this, Array.prototype.slice.call(arguments, 1));
-            });
+        if (isDefinedStr(theme.background_color)) {
+            this.el.style.backgroundColor = theme.background_color;
         }
-    }
 
-    start() {
-        let triggeredVisibleRendered = false;
-        var render = (IdleDeadline) => {
-            this.render(IdleDeadline);
+        if (isDefinedStr(theme.text_color)) {
+            this.el.style.color = theme.text_color;
+        }
 
-            if (this.viewIndex <= this.viewsLength - 1) {
-                this.renderCallbackHandle = requestIdleCallback(render);
+        if (typeof theme.line_spacing_multiplier === 'number') {
+            this.el.style.lineHeight = theme.line_spacing_multiplier;
+        }
+
+        this.el.className = 'incito';
+
+        if (this.incito.locale != null) {
+            this.el.setAttribute('lang', this.incito.locale);
+        }
+
+        this.el.innerHTML = html;
+        this.containerEl.appendChild(this.el);
+
+        this.lazyload = new LazyLoad({
+            elements_selector: '.incito--lazy',
+            data_bg: 'src',
+            threshold: 600,
+            callback_error: (el) => {
+                el.style.visibility = 'none';
+            },
+            callback_loading: (el) => {
+                if (
+                    el.tagName.toLowerCase() === 'video' &&
+                    el.getAttribute('data-controls')
+                ) {
+                    el.setAttribute('controls', 'true');
+                }
+            }
+        });
```

Comment:
Consider using `IntersectionObserver` directly instead of including an 800 line library(that is currently failing lint), something to the effect of

```js
function loadEl(el) {
    el.src = el.dataset.src;
    el.classList.remove('incito--lazy');
    if (
        el.tagName.toLowerCase() === 'video' &&
        el.getAttribute('data-controls')
    ) {
        el.setAttribute('controls', 'true');
    }
}
if ('IntersectionObserver' in window) {
    this.observer = new IntersectionObserver(
        (entries) => {
            entries.forEach((entry) => {
                if (entry.isIntersecting) {
                    loadEl(entry.target);
                    this.observer.unobserve(entry.target);
                }
            });
        },
        {root: null, rootMargin: '400px', threshold: 1.0}
    );
    this.containerEl.querySelector('.incito--lazy').forEach((lazyEl) => {
        this.observer.observe(lazyEl);
    });
} else {
    this.containerEl.querySelector('.incito--lazy').forEach((lazyEl) => {
        loadEl(lazyEl);
    });
}
```

Reply:
That's really nice. It just be a slight challenge loading everything immediately for clients (IE11) not supporting IntersectionObserver?

Reply:
Yeah but that's also what `lazyload.js` does https://github.com/shopgun/shopgun-js-sdk/blob/d80b229d744e7c647a635e2068450fa2d5a4a57b/lib/incito-browser/lazyload.js#L722

Thinking about it, an even better option is for the renderer to detect `IntersectionObserver` support and set `src` instead of `data-src` immediately, then this `} else {` is unnecessary.

Reply:
@klarstrup That's a great idea. Committed that now.